types/seat/wlr_seat_pointer.c

```diff
@@ -339,26 +340,33 @@ void wlr_seat_pointer_notify_motion(struct wlr_seat *wlr_seat, uint32_t time,
 uint32_t wlr_seat_pointer_notify_button(struct wlr_seat *wlr_seat,
 		uint32_t time, uint32_t button, enum wlr_button_state state) {
 	clock_gettime(CLOCK_MONOTONIC, &wlr_seat->last_event);
+
+	struct wlr_seat_pointer_state* pointer_state = &wlr_seat->pointer_state;
+
 	if (state == WLR_BUTTON_PRESSED) {
-		if (wlr_seat->pointer_state.button_count == 0) {
-			wlr_seat->pointer_state.grab_button = button;
-			wlr_seat->pointer_state.grab_time = time;
+		if (pointer_state->button_count == 0) {
+			pointer_state->grab_button = button;
+			pointer_state->grab_time = time;
 		}
-		wlr_seat->pointer_state.button_count++;
+		set_add(pointer_state->buttons, &pointer_state->button_count,
+			WLR_POINTER_BUTTONS_CAP, button);
+	/* } else if (state == WLR_BUTTON_RELEASED) { */
 	} else {
-		if (wlr_seat->pointer_state.button_count == 0) {
-			wlr_log(WLR_ERROR, "Corrupted seat button count");
-		} else {
-			wlr_seat->pointer_state.button_count--;
-		}
+		set_remove(pointer_state->buttons, &pointer_state->button_count,
+			WLR_POINTER_BUTTONS_CAP, button);
 	}
 
-	struct wlr_seat_pointer_grab *grab = wlr_seat->pointer_state.grab;
+
+	struct wlr_seat_pointer_grab *grab = pointer_state->grab;
 	uint32_t serial = grab->interface->button(grab, time, button, state);
 
-	if (serial && wlr_seat->pointer_state.button_count == 1 &&
+	wlr_log(WLR_DEBUG, "button_count=%"PRIu32" grab_serial=%"PRIu32" serial=%"PRIu32"",
```

Comment:
I'm not sure we want to keep this one, it'll be pretty verbose

Reply:
Yeah, maybe. The average person clicks around 5000 times a day. However, I found it fairly useful to be able to have a user action I could use to track. For me, the IPC debug logs greatly overpowered these logs (as it stands, I have to rotate my debug logs constantly or they grow to > 1.5GB a day).

The main reason I chose to leave them in was due to this being a somewhat open problem still as far as corruption goes, and I thought it could be useful in debugging future issues.